File: Data Structures & Algorithms/counting-bits/submission-4.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        std::vector<int> res{};
        res.push_back(0);
        /*
            - 1111
            - 1000
        */

        for (int i = 1; i <= n; i++) {
            int count{};
            int x = i;

            while (x) {
                count += x & 1;
                x >>= 1;
            }

            res.push_back(count);
        }

        return res;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = [0] * len(nums)

        # start from end, count curr sequence length as 1
        # check rest of the right array and track max of just this letter vs me + the right side
        # return dp[0]

        for i in range(len(nums) - 1, -1, -1):
            if i == len(nums):
                dp[i] = 1
                continue

            currLIS = 1
            for j in range(i, len(nums)):
                if nums[j] > nums[i]:
                    currLIS = max(currLIS, 1 + dp[j])

            dp[i] = currLIS

        return max(dp)